File: Raven.Client.Lightweight/Connection/AdminServerClient.cs

// -----------------------------------------------------------------------
//  <copyright file="AdminDatabaseCommands.cs" company="Hibernating Rhinos LTD">
//      Copyright (c) Hibernating Rhinos LTD. All rights reserved.
//  </copyright>
// -----------------------------------------------------------------------
using Raven.Abstractions.Data;
using Raven.Abstractions.Util;
using Raven.Client.Connection.Async;
using Raven.Client.Extensions;
using Raven.Json.Linq;

namespace Raven.Client.Connection
{
    public class AdminServerClient : IAdminDatabaseCommands, IGlobalAdminDatabaseCommands
    {
        private readonly AsyncServerClient asyncServerClient;
        private readonly AsyncAdminServerClient asyncAdminServerClient;

        public AdminServerClient(AsyncServerClient asyncServerClient, AsyncAdminServerClient asyncAdminServerClient)
        {
            this.asyncServerClient = asyncServerClient;
            this.asyncAdminServerClient = asyncAdminServerClient;
        }

        public void CreateDatabase(DatabaseDocument databaseDocument)
        {
            AsyncHelpers.RunSync(() => asyncAdminServerClient.CreateDatabaseAsync(databaseDocument));
        }

        public void DeleteDatabase(string databaseName, bool hardDelete = false)
        {
            AsyncHelpers.RunSync(() => asyncAdminServerClient.DeleteDatabaseAsync(databaseName, hardDelete));
        }

        public void EnsureDatabaseExists(string name, bool ignoreFailures = false)
        {
            AsyncHelpers.RunSync(() => asyncAdminServerClient.EnsureDatabaseExistsAsync(name, ignoreFailures));
        }

        public IDatabaseCommands Commands { get { return new ServerClient(asyncServerClient); } }

        public Operation CompactDatabase(string databaseName)
        {
            return AsyncHelpers.RunSync(() => asyncAdminServerClient.CompactDatabaseAsync(databaseName));
        }

        public void StopIndexing()
        {
            AsyncHelpers.RunSync(() => asyncAdminServerClient.StopIndexingAsync());
        }

        public void StartIndexing(int? maxNumberOfParallelIndexTasks)
        {
            AsyncHelpers.RunSync(() => asyncAdminServerClient.StartIndexingAsync(maxNumberOfParallelIndexTasks));
        }

        public Operation StartBackup(string backupLocation, DatabaseDocument databaseDocument, bool incremental, string databaseName)
        {
            return AsyncHelpers.RunSync(() => asyncAdminServerClient.StartBackupAsync(backupLocation, databaseDocument, incremental, databaseName));
        }

        public Operation StartRestore(DatabaseRestoreRequest restoreRequest)
        {
            return AsyncHelpers.RunSync(() => asyncAdminServerClient.StartRestoreAsync(restoreRequest));
        }

        public IndexingStatus GetIndexingStatus()
        {
            return AsyncHelpers.RunSync(() => asyncAdminServerClient.GetIndexingStatusAsync());
        }

        public RavenJObject GetDatabaseConfiguration()
        {
            return AsyncHelpers.RunSync(() => asyncAdminServerClient.GetDatabaseConfigurationAsync());
        }

        public BuildNumber GetBuildNumber()
        {
            return AsyncHelpers.RunSync(() => asyncAdminServerClient.GetBuildNumberAsync());
        }

        public string[] GetDatabaseNames(int pageSize, int start = 0)
        {
            return AsyncHelpers.RunSync(() => asyncAdminServerClient.GetDatabaseNamesAsync(pageSize, start));
        }

        public AdminStatistics GetStatistics()
        {
            return AsyncHelpers.RunSync(() => asyncAdminServerClient.GetStatisticsAsync());
        }
    }
}

File: Raven.Client.Java/src/main/java/net/ravendb/client/connection/AdminServerClient.java
package net.ravendb.client.connection;

import java.io.IOException;

import net.ravendb.abstractions.basic.Reference;
import net.ravendb.abstractions.closure.Function1;
import net.ravendb.abstractions.closure.Function2;
import net.ravendb.abstractions.closure.Function3;
import net.ravendb.abstractions.data.AdminStatistics;
import net.ravendb.abstractions.data.BackupRequest;
import net.ravendb.abstractions.data.DatabaseDocument;
import net.ravendb.abstractions.data.HttpMethods;
import net.ravendb.abstractions.data.RestoreRequest;
import net.ravendb.abstractions.json.linq.RavenJObject;
import net.ravendb.abstractions.json.linq.RavenJToken;
import net.ravendb.client.connection.implementation.HttpJsonRequest;
import net.ravendb.client.document.DocumentConvention;
import net.ravendb.client.extensions.MultiDatabase;
import net.ravendb.client.indexes.RavenDocumentsByEntityName;



public class AdminServerClient implements IAdminDatabaseCommands, IGlobalAdminDatabaseCommands {

  private final ServerClient innerServerClient;
  protected final AdminRequestCreator adminRequest;

  public AdminServerClient(ServerClient serverClient) {
    innerServerClient = serverClient;
    adminRequest = new AdminRequestCreator(new Function2<String, HttpMethods, HttpJsonRequest>() {

      @Override
      public HttpJsonRequest apply(String url, HttpMethods method) {
        return ((ServerClient)innerServerClient.forSystemDatabase()).createRequest(method, url);
      }
    }, new Function2<String, HttpMethods, HttpJsonRequest>() {

      @Override
      public HttpJsonRequest apply(String url, HttpMethods method) {
        return innerServerClient.createRequest(method, url);
      }
    }, new Function3<String, String, HttpMethods, HttpJsonRequest>() {

      @Override
      public HttpJsonRequest apply(String currentServerUrl, String requestUrl, HttpMethods method) {
        return innerServerClient.createReplicationAwareRequest(currentServerUrl, requestUrl, method);
      }
    });
  }

  @Override
  public void createDatabase(DatabaseDocument databaseDocument) {
    Reference<RavenJObject> docRef = new Reference<>();
    HttpJsonRequest req = adminRequest.createDatabase(databaseDocument, docRef);

    req.write(docRef.value.toString());
    req.executeRequest();
  }

  @Override
  public void deleteDatabase(String databaseName)  {
    deleteDatabase(databaseName, false);

  }
  @Override
  public void deleteDatabase(String databaseName, boolean hardDelete) {
    adminRequest.deleteDatabase(databaseName, hardDelete).executeRequest();
  }

  @Override
  public IDatabaseCommands getCommands() {
    return innerServerClient;
  }

  @Override
  public void compactDatabase(String databaseName) {
    adminRequest.compactDatabase(databaseName).executeRequest();
  }

  @Override
  public void stopIndexing() {
    innerServerClient.executeWithReplication(HttpMethods.POST, new Function1<OperationMetadata, Void>() {

      @Override
      public Void apply(OperationMetadata operationMetadata) {
        adminRequest.stopIndexing(operationMetadata.getUrl()).executeRequest();
        return null;
      }
    });
  }

  @Override
  public void startIndexing() {
    innerServerClient.executeWithReplication(HttpMethods.POST, new Function1<OperationMetadata, Void>() {

      @Override
      public Void apply(OperationMetadata operationMetadata) {
        adminRequest.startIndexing(operationMetadata.getUrl()).executeRequest();
        return null;
      }
    });
  }

  @Override
  public void startBackup(String backupLocation, DatabaseDocument databaseDocument, boolean incremental, String databaseName) {
    HttpJsonRequest request = adminRequest.startBackup(backupLocation, databaseDocument, databaseName, incremental);

    BackupRequest backupRequest = new BackupRequest();
    backupRequest.setBackupLocation(backupLocation);
    backupRequest.setDatabaseDocument(databaseDocument);

    request.write(RavenJObject.fromObject(backupRequest).toString());
    request.executeRequest();
  }

  @Override
  public void startRestore(RestoreRequest restoreRequest) {
    HttpJsonRequest request = adminRequest.createRestoreRequest();
    request.write(RavenJObject.fromObject(restoreRequest).toString());
    request.executeRequest();
  }


  @Override
  public String getIndexingStatus() {
    return innerServerClient.executeWithReplication(HttpMethods.GET, new Function1<OperationMetadata, String>() {

      @Override
      public String apply(OperationMetadata operationMetadata) {
        RavenJToken result = adminRequest.indexingStatus(operationMetadata.getUrl()).readResponseJson();
        return result.value(String.class, "IndexingStatus");
      }
    });
  }

  @Override
  public AdminStatistics getStatistics() {
    try {
      RavenJObject json = (RavenJObject)adminRequest.adminStats().readResponseJson();
      return innerServerClient.convention.createSerializer().readValue(json.toString(), AdminStatistics.class);
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
  }

  @Override
  public void ensureDatabaseExists(String name, boolean ignoreFailures) {

    ServerClient serverClient = (ServerClient) innerServerClient.forSystemDatabase();

    try (AutoCloseable readFromMaster = serverClient.forceReadFromMaster()) {
      DatabaseDocument doc = MultiDatabase.createDatabaseDocument(name);

      try {
        if (serverClient.get(doc.getId()) != null) {
          return;
        }
        serverClient.getGlobalAdmin().createDatabase(doc);
      } catch (Exception e) {
        if (!ignoreFailures) {
          throw new RuntimeException(e);
        }
      }

      try {
        new RavenDocumentsByEntityName().execute(serverClient.forDatabase(name), new DocumentConvention());
      } catch (Exception e) {
        // we really don't care if this fails, and it might, if the user doesn't have permissions on the new db
      }

    } catch (Exception e) {
        // we really don't care if this fails, and it might, if the user doesn't have permissions on the new db
    }
  }

  @Override
  public void ensureDatabaseExists(String name) {
    ensureDatabaseExists(name, false);
  }


}
